### ISSUE TITLE
generate_waymo_tf.py hangs 

### ISSUE THREAD

**Issue by JWHennessey** [tensorflow/lingvo#214]
Hi, 

Thanks for open sourcing such a great library. 

I am trying to convert some of the Waymo data tfrecord data into the tf-example format used my lingvo. The data hosted at waymo_open_dataset_v_1_0_0_tf_example_lingvo/v.1.0.0/ doesn't include the data for the domain adaption subset as far as I can tell. 

I've tried running the generate_waymo_tf.py script both with the DataflowRunner and DirectRunner. Both seem to end up hanging and never generating any outputs. 

Are there any suggestions on how to debug this? 

Here is how I run the script which should be trying to process only a single segment.
```
python3 /tmp/lingvo/lingvo/tasks/car/waymo/tools/generate_waymo_tf.py \
  --input_file_pattern=gs://{bucket}/data/domain_adaptation/validation/segment-10302420246752500683_5520_000_5540_000* \
  --output_filebase=gs://{bucket}/tf_example_lingvo/domain_adaptation/validation/output -- \
  --project='{project}' \
  --temp_location=gs://{bucket}/tmp/ \
  --region 'europe-west4' \
  --zone 'europe-west4-c' \
  --num_workers 2 \
  --setup_file /tmp/lingvo/lingvo/tasks/car/waymo/tools/setup.py  
```

The outputs:

```
2020-05-08 16:06:49.830048: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer.so.6'; dlerror: libnvinfer.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:49.830188: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer_plugin.so.6'; dlerror: libnvinfer_plugin.so.6: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:49.830222: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:30] Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
WARNING:absl:Lingvo does not support eager execution yet. Please disable eager execution with tf.compat.v1.disable_eager_execution() or proceed at your own risk.
I0508 16:06:50.560759 139933516105536 auth.py:108] Setting socket default timeout to 60 seconds.
I0508 16:06:50.560907 139933516105536 auth.py:111] socket default timeout is 60.0 seconds.
I0508 16:06:50.567577 139933516105536 gcsio.py:480] Starting the size estimation of the input
I0508 16:06:50.568466 139933516105536 transport.py:151] Attempting refresh to obtain initial access_token
I0508 16:06:50.639814 139933516105536 gcsio.py:496] Finished listing 1 files in 0.07223939895629883 seconds.
I0508 16:06:50.640702 139933516105536 pipeline.py:180] Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
I0508 16:06:51.375959 139933516105536 fn_api_runner_transforms.py:548] ==================== <function annotate_downstream_side_inputs at 0x7f445eff9048> ====================
I0508 16:06:51.376795 139933516105536 fn_api_runner_transforms.py:548] ==================== <function fix_side_input_pcoll_coders at 0x7f445eff9158> ====================
I0508 16:06:51.377469 139933516105536 fn_api_runner_transforms.py:548] ==================== <function lift_combiners at 0x7f445eff91e0> ====================
I0508 16:06:51.377765 139933516105536 fn_api_runner_transforms.py:548] ==================== <function expand_sdf at 0x7f445eff9268> ====================
I0508 16:06:51.378669 139933516105536 fn_api_runner_transforms.py:548] ==================== <function expand_gbk at 0x7f445eff92f0> ====================
I0508 16:06:51.379250 139933516105536 fn_api_runner_transforms.py:548] ==================== <function sink_flattens at 0x7f445eff9400> ====================
I0508 16:06:51.379594 139933516105536 fn_api_runner_transforms.py:548] ==================== <function greedily_fuse at 0x7f445eff9488> ====================
I0508 16:06:51.381523 139933516105536 fn_api_runner_transforms.py:548] ==================== <function read_to_impulse at 0x7f445eff9510> ====================
I0508 16:06:51.381770 139933516105536 fn_api_runner_transforms.py:548] ==================== <function impulse_to_input at 0x7f445eff9598> ====================
I0508 16:06:51.382106 139933516105536 fn_api_runner_transforms.py:548] ==================== <function inject_timer_pcollections at 0x7f445eff9730> ====================
I0508 16:06:51.382540 139933516105536 fn_api_runner_transforms.py:548] ==================== <function sort_stages at 0x7f445eff97b8> ====================
I0508 16:06:51.382743 139933516105536 fn_api_runner_transforms.py:548] ==================== <function window_pcollection_coders at 0x7f445eff9840> ====================
I0508 16:06:51.384798 139933516105536 statecache.py:154] Creating state cache with size 100
I0508 16:06:51.386486 139933516105536 fn_api_runner.py:2011] Created Worker handler <apache_beam.runners.portability.fn_api_runner.EmbeddedWorkerHandler object at 0x7f445ea61898> for environment urn: "beam:env:embedded_python:v1"

I0508 16:06:51.386961 139933516105536 fn_api_runner.py:974] Running (((ref_AppliedPTransform_Read/Read/_SDFBoundedSourceWrapper/Impulse_5)+(Read/Read/_SDFBoundedSourceWrapper/ParDo(SDFBoundedSourceDoFn)/PairWithRestriction))+(Read/Read/_SDFBoundedSourceWrapper/ParDo(SDFBoundedSourceDoFn)/SplitAndSizeRestriction))+(ref_PCollection_PCollection_1_split/Write)
I0508 16:06:51.425244 139933369054976 gcsio.py:480] Starting the size estimation of the input
I0508 16:06:51.478516 139933369054976 gcsio.py:496] Finished listing 1 files in 0.05326652526855469 seconds.
I0508 16:06:51.480220 139933369054976 gcsio.py:480] Starting the size estimation of the input
I0508 16:06:51.527962 139933369054976 gcsio.py:496] Finished listing 1 files in 0.04774069786071777 seconds.
I0508 16:06:51.528657 139933369054976 gcsio.py:480] Starting the size estimation of the input
I0508 16:06:51.575519 139933369054976 gcsio.py:496] Finished listing 1 files in 0.04685568809509277 seconds.
I0508 16:06:51.604114 139933516105536 fn_api_runner.py:974] Running (((((ref_AppliedPTransform_Write/Write/WriteImpl/DoOnce/Impulse_12)+(ref_AppliedPTransform_Write/Write/WriteImpl/DoOnce/FlatMap(<lambda at core.py:2639>)_13))+(ref_AppliedPTransform_Write/Write/WriteImpl/DoOnce/Map(decode)_15))+(ref_AppliedPTransform_Write/Write/WriteImpl/InitializeWrite_16))+(ref_PCollection_PCollection_6/Write))+(ref_PCollection_PCollection_7/Write)
I0508 16:06:51.626923 139933516105536 fn_api_runner.py:974] Running ((((((ref_PCollection_PCollection_1_split/Read)+(Read/Read/_SDFBoundedSourceWrapper/ParDo(SDFBoundedSourceDoFn)/Process))+(ref_AppliedPTransform_ConvertToTFExample_7))+(ref_AppliedPTransform_Write/Write/WriteImpl/WriteBundles_17))+(ref_AppliedPTransform_Write/Write/WriteImpl/Pair_18))+(ref_AppliedPTransform_Write/Write/WriteImpl/WindowInto(WindowIntoFn)_19))+(Write/Write/WriteImpl/GroupByKey/Write)
W0508 16:06:52.151952 139933369054976 tfrecordio.py:63] Couldn't find python-snappy so the implementation of _TFRecordUtil._masked_crc32c is not as fast as it could be.
2020-05-08 16:06:53.771899: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2020-05-08 16:06:54.350836: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-05-08 16:06:54.351292: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1555] Found device 0 with properties: 
pciBusID: 0000:00:04.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.75GiB deviceMemoryBandwidth: 836.37GiB/s
2020-05-08 16:06:54.351615: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.351855: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.352015: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.352181: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.352364: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.352548: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2020-05-08 16:06:54.356941: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2020-05-08 16:06:54.356972: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1592] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2020-05-08 16:06:54.357462: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-05-08 16:06:54.363119: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2000175000 Hz
2020-05-08 16:06:54.363433: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7f4450fa39b0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-05-08 16:06:54.363463: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2020-05-08 16:06:54.460026: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-05-08 16:06:54.460613: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7f445101f5a0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2020-05-08 16:06:54.460666: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Tesla V100-SXM2-16GB, Compute Capability 7.0
2020-05-08 16:06:54.460823: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1096] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-05-08 16:06:54.460847: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102]
```

Any help much appreciated. Thanks!

System Config
```
-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.104      Driver Version: 410.104      CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   36C    P0    38W / 300W |    417MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

**Comment by JWHennessey** [comment]
I found that installing the dependences on the VM and not using the docker image provided didn't have the same issue. I assume the problem was related to cuda driver versions on the host machine and docker container.

**Comment by vrv** [comment]
Hi @JWHennessey, thanks for the interest and for working through these issues.

Running beam conversion on the entire datasets can be expensive for some users, so we're going to try to have Waymo host the updated versions of the dataset in TFRecord soon, including the domain adaptation dataset.  Let us know if you successfully get your pipeline finished though, it's nice to have confirmation that the tool works for other users :)

**Comment by JWHennessey** [comment]
Hi @vrv, I will do. I haven't been able to get lingvo / starnet training on the sample dataset yet let alone the one I have converted. I will confirm if I ever manage to get it to train! :)

**Comment by vrv** [comment]
Hi @JWHennessey, not sure if it helps you still, but we finally got the files for the latest version of the data hosted by Waymo: 

gsutil ls gs://waymo_open_dataset_v_1_0_0_tf_example_lingvo/v.1.2.0

Let us know if that works for you!

**Comment by JWHennessey** [comment]
Thanks. I've converted them now but good to have as a reference incase the conversion is wrong. Much appreciated.

**Comment by JWHennessey** [comment]
One question. The domain_adaptation folder inside bucket seems to have `test*.tft*` and `test_location*tfr*`. Do you know the difference between them? 

I'd guess  `test*.tft*` is the test set for the submission but would be good to confirm. Thanks

**Comment by bcaine** [comment]
Hey @JWHennessey, the short answer is you want to use both.

They're different subsets of the overall test set. This will also exist in the original Waymo Open Dataset (a test-* and a test_location-*) if I remember correctly. 

Let me know if you happen to also be using lingvo to generate your test set predictions and need some help figuring out pull from multiple data paths.

**Comment by JWHennessey** [comment]
Thanks @bcaine for that. Previously I converted all the segments in `waymo_open_dataset_v_1_2_0_individual_files/domain_adaptation/testing` to TF Example. Will I have both of these subsets already? 

To be on the safe side I may use these ones. We will being using Lingvo. So I guess something like `p.file_datasource.file_pattern = test*.tfr-0????-of-01000` should do the trick in `params/waymo.py`?

**Comment by vrv** [comment]
You probably don't have those subsets broken out, I think that's probably from the internal directory structure we used from waymo which isn't structured that way on GCS, though it's the same data.

I think test + test_location is the equivalent, so yes, that should probably do the trick to capture all of the test files in our version.
